Привет! Меня зовут Александр Арасланов, я провожу ревью твоей работы. Предалагю общаться на «ты», если не против.

В работе ты увидишь комментарии, помеченные разным цветом.

<div class="alert alert-danger"> ❌
Красным цветом помечены моменты, которые необходимо доработать, чтобы проект был принят. Уверен, что у тебя все получится. Буду ждать твою работу на повторное ревью. 
</div>

<div class="alert alert-success"> ✅
Зеленым цветом помечены оброзцовые решения. Которые можно использовать в дальнейшем.
</div>

<div class="alert alert-warning"> ⚠️
Желтым цветом отмечено то, что ты можешь улучшить или доработать. Можешь учесть это при выполнении следующих работ или доработать проект сейчас, но это не обязательно.
</div>


Будет комфортнее работать над проектом вместе, если ты будешь **писать об изменениях** в проекте по моим рекомендациям. Для этого выбери для своих комментариев 

<div class="alert alert-block alert-info">заметный цвет</div>

Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее.

Проект снова отправился не тот. Пожалуйста, проверь все внимательно перед отправкой. В проекте отсутсвует 3 пункт. 
<div class="alert alert-block alert-info">Добавила третий пункт</div>


# Исследование надежности заемщиков

Для кредитного отдела банка мы проводим исследование надежности заемщиков. 
На основе статистики платежеспособности клиентов мы должны узнать, какие характеристики клиентов влияют на возврат кредита в срок.

**Цель исследования** - проверить четыре гипотезы: 
1. Семейное положение клиента влияет на факт погашения кредита в срок.
2. Количество детей у клиента влияет на факт погашения кредита в срок.
3. Уровень дохода клиента влияет на факт погашения кредита в срок.
4. Цели кредита по-разному влияют на факт погашения кредита в срок.

**Ход исследования**
Данные о платежеспособности клиентов мы получили из файла `data.csv`. О качестве данныз нам ничего неизвестно, поэтому перед проверкой гипотез понадобится обзор данных.

Мы проверим данные на ошибки и оценим их влияние на исследование, затем, на этапе предобработки мы поищем возможность исправить самые критичные ошибки данных.

С помощью перекрестных таблиц мы оценим влияние различных факторов на факт погашения крелита в срок. 

Таким образом, решение пройдет в три этапа:


## Обзор данных

Составим представление о данных. Импортируем специальную библиотеку pandas для обработки и анализа данных. 

In [57]:
import pandas as pd # импорт библиотеки pandas

Прочитаем файл `data.csv` и сохраним его в переменной `df`:

In [58]:
# чтение файла с данными и сохранение в df
try:
    # адрес в локальном окружении
    df = pd.read_csv('data.csv')
except:
    # адрес в среде 'practicum'
    df = pd.read_csv('/datasets/data.csv')  

Выведем на экран первые десять строк таблицы

In [59]:
display(df.head(10)) # получение первых 10 строк таблицы df

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим общую информацию о таблице

In [60]:
df.info() # получение общей информации о данных в таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Итак, в таблице 11 столбцов. 

Согласно документации к данным:
* `children` — количество детей в семье, тип данных - int64;
* `days_employed` — общий трудовой стаж в днях, тип данных - float64;
* `dob_years` — возраст клиента в годах, тип данных - int64;
* `education` — уровень образования клиента, тип данных - object;
* `education_id` — идентификатор уровня образования, тип данных - int64;
* `family_status` — семейное положение, тип данных - object;
* `family_status_id` — идентификатор семейного положения, тип данных - int64;
* `gender` — пол клиента, тип данных - object;
* `income_type` — тип занятости, тип данных - object;
* `debt` — имел ли задолженность по возврату кредитов, тип данных - int64;
* `total_income` — ежемесячный доход, тип данных - float64;
* `purpose` — цель получения кредита, тип данных - object;

В названиях колонок нет нарушений.

Количество значений в столбцах различается. Значит, есть пропущенные значения. Конкретно - в столбцах `days_employed` и `total_income`.

**Выводы**

В каждой строке таблицы - данные о заемщике. Часть колонок описывает социально-демографические характеристики характеристики - возраст, пол, семейное положение, количество детей. Другие колонки характеризуют образование и занятость заемщика - уровень образования, стаж, тип занятости, доход. Оставшиеся описывают характеристики кредита - цель получения и наличие задолженности по возврату. 

Предварительно можно утверждать что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных

Исключим пропуски данных и дубликатов, проверим данные на наличие аномалий и избавимся от них. Приведем данные к необходимым типам.
Для удобства работы декомпозируем исходный датафрейм, подготовив ряд словарей. 
Подготовим категории для столбцов 'цель кредита' и 'уровень дохода'.

### Заполнение пропусков

В некоторых столбцах есть приопущенные значения. Оценим их количество. 

In [61]:
df.isna().sum() # подсчёт пропусков

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Итак, пропуски есть в двух столбцах. Один из них — `days_employed`, другой - `total_income`. Интересно, что пропусков одинаковое количество. 
Проанализируем эти столбцы. Запросим первые 10 строк с пропущенными значениями.

In [62]:
# просмотр 10 первых строк с пропущенными значениями
display(df[df['total_income'].isna()].head(10))

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Пропущенные значения не идут подряд, они не связаны ни с особенностями клиентов (возрастом, занятостью),ни с типом кредита. Однако пропуски в столбце `days_employed` соответствуют пропускам в столбце `total_income`. Проверим, так ли это во всей таблице.

In [63]:
# Количество строк, где одновременно пропущенны значения в столбцах 'days_employed' и 'total_income'
df_isna = len(df.loc[df['days_employed'].isna() & df['total_income'].isna()])
print(f'Количество строк, где пропущенны значения одновременно в обоих столбцах {df_isna}')

# Доля этого количества от общего количества наблюдений
na_share_2 = df.isna().mean()
print(f'Доля строк с пропущенными значениями от общего числа строк в таблице составляет {na_share_2}')


Количество строк, где пропущенны значения одновременно в обоих столбцах 2174
Доля строк с пропущенными значениями от общего числа строк в таблице составляет children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64


Пропущенные значения составляют примерно 10% в каждом из столбцов с пропусками. Это немало.
Только столбцы `days_employed`и `total_income` в датафрейме имеют формат `float64`. Совпадение пропусков в этих столбцах говорит о том, что может быть допущена техническая ошибка при выгрузке или обработке данных. 
В реальном проекте необходимо было бы обратиться к источнику данных, найти и исправить техническую ошибку.

У нас такой возможности нет, поэтому постараемся заполнить отсутствующие данные типичными значениями выборки. 
Начнем со столбца `total_income`. Для заполнения пропусков можно использовать типичные значения выборки - среднее арифметическое или медиану. Рассчитаем оба этих показателя.

In [64]:
mean_total_income = df['total_income'].mean() # расчет среднего арифметического по столбцу
median_total_income = df['total_income'].median()  # расчет медианы по столбцу

print(f'Среднее арифметическое значение по столбцу \'Ежемесячный доход\' - {mean_total_income:.2f}' )
print(f'Медианное значение по столбцу \'Ежемесячный доход\' - {median_total_income:.2f}' )

Среднее арифметическое значение по столбцу 'Ежемесячный доход' - 167422.30
Медианное значение по столбцу 'Ежемесячный доход' - 145017.94


Мы видим, что среднее арифметическое значение значительно выше медианного. Это объясняется тем, что медианное значение менее чувствительно к 'выбросам' - очень большим и очень маленьким значениям в выборке, которые могут сдвигать среднее. В нашем случае малое количество заемщиков с очень высокими доходами влияет на среднее арифметическое, повышая его. На медиану 'выбросы' практически не влияют.

Поэтому лучшим решением будет выбрать для замены медианное значение. 

In [65]:
# Заполнение пропущенных значений с помощью метода fillna()
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))# Проверка на отсутствие пропущенных значений
total_income_isna = df['total_income'].isna().sum()
print(f'Количество строк, где пропущенны значения в столбцe \'Ежемесячный доход\' - {total_income_isna}')

Количество строк, где пропущенны значения в столбцe 'Ежемесячный доход' - 0


Мы заменили все пропущенные значения в столбце `total_income` на медианные. Перейдем к столбцу `days_employed`

### Проверка данных на аномалии и исправления.

Проанализируем столбец `days_employed`. Рассчитаем минимальное, максимальное, медианное и среднее значение по столбцу.

In [66]:
max_days_employed = df['days_employed'].max() # расчет максимального значения по столбцу
min_days_employed = df['days_employed'].min() # расчет минимального значения по столбцу
mean_days_employed = df['days_employed'].mean() # расчет среднего арифметического по столбцу
median_days_employed = df['days_employed'].median()  # расчет медианы по столбцу

print(f'Максимальное арифметическое значение по столбцу \'Стаж в днях\' - {max_days_employed:.2f}' )
print(f'Минимальное значение по столбцу \'Стаж в днях\' - {min_days_employed:.2f}' )
print(f'Среднее арифметическое значение по столбцу \'Стаж в днях\' - {mean_days_employed:.2f}' )
print(f'Медианное значение по столбцу \'Стаж в днях\' - {median_days_employed:.2f}' )

Максимальное арифметическое значение по столбцу 'Стаж в днях' - 401755.40
Минимальное значение по столбцу 'Стаж в днях' - -18388.95
Среднее арифметическое значение по столбцу 'Стаж в днях' - 63046.50
Медианное значение по столбцу 'Стаж в днях' - -1203.37


В данных встречаюся артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. В столбце `days_employed` это отрицательные значения, слишком большые положительные значения. 
Для реальных данных это нормально, но надо их учесть.
Просмотрим начало датафрейма, попробуем увидеть ошибки невооруженным взглядом

In [67]:
df.head(15) # показ первых 15 строк датафрейма.

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  118514.486412                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Проверим данные, сгруппировав по типу дохода, возможно эта характеристика влияет на аномальные значения.

In [68]:
'''Группировка стажа по типу дохода
Cреднее значение делится на 365, чтобы получить среднее количество лет стажа по категории '''
df['days_employed'].groupby(by = df['income_type']).mean()/365 

income_type
безработный        1003.873021
в декрете            -9.032219
госслужащий          -9.314786
компаньон            -5.784998
пенсионер          1000.009565
предприниматель      -1.426981
сотрудник            -6.373970
студент              -1.585621
Name: days_employed, dtype: float64

Итак, наша гипотеза нашла подтверждение - у безработных и пенсионеров средний стаж выше 1000 лет, это значит, что в данных какая-то ошибка. Возможная причина - другой источник формирования информации о стаже, другой формат данных, техническая ошибка при заполнении.
Проверим данные на наличие аномалий в других группах по источнику дохода:

In [69]:
# Проверка, что в группах, кроме пенсионеров и безработных, нет слишком больших абсолютных значений стажа 
df.loc[(df['income_type'] != 'пенсионер') 
       & (df['income_type'] != 'безработный') 
       & (abs(df['days_employed']) > 25000)] # 25000 дней примерно соответствует 70 годам

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

В других группах слишком больших абсолютных значений стажа нет. Но в них другая проблема - отрицательная величина этого показателя. 
Избавимся от отрицательных значений в столбце `days_employed`, заменив их на абсолютные значения.

In [70]:
# Замена отрицательных значений в столбце 'days_employed'
# Все значения заменяются на абсолютные
df['days_employed'] = abs(df['days_employed'])
# Контроль - просмотр первых пяти строк датафрейма
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Возможно, что стаж безработных и пенсионеров исчисляется не в днях, а в часах. 

In [71]:
'''Проверка стажа безработных и пенсионеров
Сначала показатель делится на 24, чтобы получить значение в днях
Затем среднее значение делится на 365, чтобы получить среднее количество лет стажа'''
min_elder_unemp_years = df.loc[(df['income_type'] == 'пенсионер')
                         | (df['income_type'] =='безработный') ]['days_employed'].min()/24/365
max_elder_unemp_years = df.loc[(df['income_type'] == 'пенсионер')
                         | (df['income_type'] =='безработный') ]['days_employed'].max()/24/365
print(f'Минимальное количество лет стажа пенсионеров и безработных после преобразования {min_elder_unemp_years:.2f}')
print(f'Максимальное количество лет стажа пенсионеров и безработных после преобразования {max_elder_unemp_years:.2f}')

Минимальное количество лет стажа пенсионеров и безработных после преобразования 37.53
Максимальное количество лет стажа пенсионеров и безработных после преобразования 45.86


Наша гипотеза подтверждается, после деления на 24 стаж в группах пенсионеров и безработных выглядит реалистично. Значит, сейчас данные по этим группам в таблице указаны не в днях, а в часах и необходимо преобразовать значения в колонках.

In [72]:
# Проведение преобразования: 
# для категорий 'пенсионер' и 'безработный' все значения в столбце days_employed делятся на 24.
df.loc[(df['income_type'] == 'пенсионер') 
       | (df['income_type'] == 'безработный'), 'days_employed'] = df['days_employed'] / 24
# Проверка успешности преобразования
df[df['income_type'] == 'пенсионер'].head()

children  days_employed  dob_years education  education_id  \
4          0   14177.753002         53   среднее             1   
12         0            NaN         65   среднее             1   
18         0   16678.380705         53   среднее             1   
24         1   14106.331371         57   среднее             1   
25         0   15147.853723         67   среднее             1   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
12       гражданский брак                 1      M   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
12  118514.486412                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости

Проверим другие колонки. Начнем с колонки 'дети'.

In [73]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В столбце 'дети' встречаются неверные значения: 20 и -1, что скорее всего соответствует 2 и 1 ребенку. 

In [74]:
df.loc[df['children'] == 20, 'children'] = 2 
df.loc[df['children'] == -1, 'children'] = 1 
df['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Со столбцом 'дети' все в порядке. Проверим столбец 'возраст'. Здесь в наличии заемщики младше 18 лет. Это техническая ошибка, так как кредит можно выдавать только совершеннолетним. В реальности необходимо было бы обратиться к источнику данных и поработать с ним. Мы же заменим эти значения на типичные - медианные по группам занятости.

In [75]:
df[df['dob_years'] < 18].groupby('income_type')['income_type'].count()


income_type
госслужащий     6
компаньон      20
пенсионер      20
сотрудник      55
Name: income_type, dtype: int64

In [76]:
median_age_state = df.loc[df['income_type'] == 'госслужащий']['dob_years'].median()
median_age_comp = df.loc[df['income_type'] == 'компаньон']['dob_years'].median()
median_age_elder = df.loc[df['income_type'] == 'пенсионер']['dob_years'].median()
median_age_empl = df.loc[df['income_type'] == 'сотрудник']['dob_years'].median()

df.loc[(df['dob_years'] < 18) & (df['income_type'] == 'госслужащий'), 'dob_years'] = median_age_state
df.loc[(df['dob_years'] < 18) & (df['income_type'] == 'компаньон'), 'dob_years'] = median_age_comp
df.loc[(df['dob_years'] < 18) & (df['income_type'] == 'пенсионер'), 'dob_years'] = median_age_elder
df.loc[(df['dob_years'] < 18) & (df['income_type'] == 'сотрудник'), 'dob_years'] = median_age_empl

df[df['dob_years'] < 18]['dob_years'].count()

0

Теперь необходимо заполнить данные по стажу. Понятно, что у студентов и госслужащий средний стаж будет отличаться, поэтому заполним отсутствующие данные по стажу медианными по группам занятости. 
Медиану мы выбираем в качестве типичного показателя, так как она менее восприимчива к 'выбросам' в данных.

Перед тем, как использовать медиану необходимо убедится, что группы достаточно наполнены. 

In [77]:
df['income_type'].groupby(df['income_type']).count() # группировка клиентов по виду дохода

income_type
безработный            2
в декрете              1
госслужащий         1459
компаньон           5085
пенсионер           3856
предприниматель        2
сотрудник          11119
студент                1
Name: income_type, dtype: int64

Запросим малонаполненные группы: безработные, в декрете, предприниматель и студент.

In [78]:
# Вывести сроки датафрейма с редко встречающимися типами занятости
df.loc[(df['income_type'] == 'безработный') 
       | (df['income_type'] == 'в декрете') 
       | (df['income_type'] == 'предприниматель') 
       | (df['income_type'] == 'студент')] 

children  days_employed  dob_years education  education_id  \
3133          1   14063.519451       31.0   среднее             1   
5936          0            NaN       58.0    высшее             0   
9410          0     578.751554       22.0    высшее             0   
14798         0   16470.951611       45.0    Высшее             0   
18697         0     520.848083       27.0    высшее             0   
20845         2    3296.759962       39.0   СРЕДНЕЕ             1   

               family_status  family_status_id gender      income_type  debt  \
3133         женат / замужем                 0      M      безработный     1   
5936         женат / замужем                 0      M  предприниматель     0   
9410   Не женат / не замужем                 4      M          студент     0   
14798       гражданский брак                 1      F      безработный     0   
18697       гражданский брак                 1      F  предприниматель     0   
20845        женат / замужем                 0      F        в декрете     1   

        total_income                                 purpose  
3133    59956.991984                 покупка жилья для сдачи  
5936   499163.144947              покупка жилой недвижимости  
9410    98201.625314  строительство собственной недвижимости  
14798  202722.511368                            ремонт жилью  
18697  499163.144947                   на проведение свадьбы  
20845   53829.130729                              автомобиль

Пропущенные значения встречаются только в категории 'предприниматель'. 
Теперь проверим наполненные категории: госслужащий, компаньон, пенсионер, сотрудник. 

In [79]:
# Подсчет количества пропущенных значений в категории 'госслужащие'
isna_state = df.loc[df['income_type'] == 'госслужащий']['days_employed'].isna().sum()
isna_comp = df.loc[df['income_type'] == 'компаньон']['days_employed'].isna().sum() 
isna_elder = df.loc[df['income_type'] == 'пенсионер']['days_employed'].isna().sum() 
isna_empl = df.loc[df['income_type'] == 'сотрудник']['days_employed'].isna().sum() 

print(f'Пропущенных значений в группе \'госслужащий\' - {isna_state}')
print(f'Медианный стаж для группы \'компаньон\' - {isna_comp}')
print(f'Медианный стаж для группы \'пенсионер\' - {isna_elder}')
print(f'Медианный стаж для группы \'сотрудник\' - {isna_empl}')

Пропущенных значений в группе 'госслужащий' - 147
Медианный стаж для группы 'компаньон' - 508
Медианный стаж для группы 'пенсионер' - 413
Медианный стаж для группы 'сотрудник' - 1105


Среди госслужащих нет пропущенных значений. Для остальных категорий рассчитаем медианные значения.

In [80]:
# Группировка данных по типу занятости, получение медианного значения
median_comp = df.loc[df['income_type'] == 'компаньон']['days_employed'].median()
median_elder = df.loc[df['income_type'] == 'пенсионер']['days_employed'].median()
median_empl = df.loc[df['income_type'] == 'сотрудник']['days_employed'].median()
median_state = df.loc[df['income_type'] == 'госслужащий']['days_employed'].median()
print(f'Медианный стаж для группы \'компаньон\' - {median_comp:.0f}  дней')
print(f'Медианный стаж для группы \'пенсионер\' - {median_elder:.0f} дней')
print(f'Медианный стаж для группы \'сотрудник\' - {median_empl:.0f} дней')
print(f'Медианный стаж для группы \'госслужащий\' - {median_state:.0f} дней')


Медианный стаж для группы 'компаньон' - 1547  дней
Медианный стаж для группы 'пенсионер' - 15217 дней
Медианный стаж для группы 'сотрудник' - 1574 дней
Медианный стаж для группы 'госслужащий' - 2689 дней


Теперь заполним пропуски в days_employed медианными значениями по этому столбцу. Для группы 'предприниматель' используем медианное значение по группе 'компаньон' - логически это наиболее близкая категория. 

In [81]:
# Замена значений стажа медианными по типу дохода. 
df.loc[df['income_type'] == 'компаньон','days_employed'] = df.loc[df['income_type'] == 'компаньон','days_employed'].fillna(median_comp)
df.loc[df['income_type'] == 'предприниматель','days_employed'] = df.loc[df['income_type'] == 'предприниматель','days_employed'].fillna(median_comp)
df.loc[df['income_type'] == 'пенсионер','days_employed'] = df.loc[df['income_type'] == 'пенсионер','days_employed'].fillna(median_elder)
df.loc[df['income_type'] == 'сотрудник','days_employed'] = df.loc[df['income_type'] == 'сотрудник','days_employed'].fillna(median_empl)
df.loc[df['income_type'] == 'госслужащий','days_employed'] = df.loc[df['income_type'] == 'госслужащий','days_employed'].fillna(median_state)


Убедимся, что в таблице не осталось пропусков. Для этого ещё раз посчитаем пропущенные значения.

In [82]:
# Запрос пропущенных значений по всем столбцам.
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Пропущенных значений нет, можем двигаться дальше.

### Изменение типов данных.

Заменим вещественный тип данных в столбце total_income на целочисленный, например, с помощью метода astype()

In [83]:
# Замена вещественного типа данных в столбце total_income на целочисленный с помощью метода .astype()
df['total_income'] = df['total_income'].astype('int64')
# Проверка успешности замены - запрос информации обо всех столбцах.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  float64
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Замена произведена успешно.

### Удаление дубликатов.

Если в данных присутствуют строки-дубликаты, удалим их. 

In [84]:
df.duplicated().sum() # подсчёт явных дубликатов

54

Явные дубликаты есть, их 54. Вызовем специальный метод `pandas`, чтобы удалить их:

In [85]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True) 

Ещё раз посчитаем явные дубликаты в таблице — убедимся, что полностью от них избавились:

In [86]:
df.duplicated().sum() # проверка на отсутствие дубликатов

0

Теперь явных дубликатов нет. Обработаем неявные дубликаты. В столбце education есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к одному регистру.

In [87]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь все значения столбца "образование" представлены единообразно. 
Проверим остальные столбцы. Столбцы "дети" и "возраст" мы проверяли в ходе анализа аномалий. С ним все в порядке. Проверим столбец "семейное положение".

In [88]:
# Запрос значений в столбце 'family_status'
df['family_status'].value_counts()

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

Дубликатов нет, однако формат не везде соблюден - регистр варианта "Не женат / не замужем" отличается от остальных вариантов. Приведем к единообразию. 

In [89]:
# Преобразование столбца в нижний регистр с помощью метода str.lower()
df['family_status'] = df['family_status'].str.lower()
# Проверка - запрос значений в столбце 'family_status'
df['family_status'].value_counts()    

женат / замужем          12344
гражданский брак          4163
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

Теперь проверим столбец "Пол".

In [90]:
df['gender'].value_counts()  

F      14189
M       7281
XNA        1
Name: gender, dtype: int64

Одна из строк имеет значение XNA. Если бы это было важно для анализа, мы бы удалили данную строку. Так как пол не важен в наших рассчетах, оставим так, тем более что третий пол сейчас активно используется во многих анкетах.

In [91]:
df['income_type'].value_counts()  

сотрудник          11091
компаньон           5080
пенсионер           3837
госслужащий         1457
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

Категория "предприниматель" логически соответствует категории "компаньон", можно их объединить. 

In [92]:
df.loc[df['income_type'] == 'предприниматель','income_type'] = 'компаньон'
df['income_type'].value_counts()  

сотрудник      11091
компаньон       5082
пенсионер       3837
госслужащий     1457
безработный        2
в декрете          1
студент            1
Name: income_type, dtype: int64

Теперь в типах дохода представлены только различные варианты. Проверим столбец "задолженность".

In [93]:
df['debt'].value_counts()  

0    19730
1     1741
Name: debt, dtype: int64

Хорошо, что хотя бы этот столбец не имеет аномалий и дубликатов. Он критически важен для нашего исследования.

Теперь еще раз проверим на явные дубликаты, чтобы убедиться, что в ходе обработки они не появились.

In [94]:
df.duplicated().sum() # подсчёт явных дубликатов

17

Удалим появившиеся дубликаты и проверим, что дубликатов в данных не осталось.

In [95]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True) 
df.duplicated().sum() # проверка на отсутствие дубликатов

0

Мы удалили дубликаты в наших данных. Для этого мы использовали следующие методы:
1. duplicated() и drop_duplicates() для обработки явных дубликатов - их поиска и удаления. 
чтобы не сомневаться при удалении явных дубликатов, можно в таблицу добавить уникальный ключ клиента. 
2. value_counts() для поиска похожих значений перед ручным удалением дубликатов.
3. str.lower() для приведения к единому регистру (в нашем случае - к нижнему)
4. df.loc(с условием) - для замены по значению.
Выбирали метод в зависимости от природы дубликатов.

Возможные причины появления дубликатов: ошибки в процессе ввода данных - оператор может случайно ввести информацию несколько раз, отсутствие словарей и разные стандарты в разных источниках данных. 

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

Создадим два новых датафрейма:
в первом словаре *education_dict* каждому уникальному значению из education соответствует уникальное значение education_id.

In [96]:
education_dict = df[['education', 'education_id']]
education_dict = education_dict.drop_duplicates()
display(education_dict)


education  education_id
0                  высшее             0
1                 среднее             1
13    неоконченное высшее             2
31              начальное             3
2962       ученая степень             4

Во втором словаре *family_status_dict* каждому уникальному значению из family_status соответствует уникальное значение family_status_id

In [97]:
family_status_dict = df[['family_status', 'family_status_id']]
family_status_dict = family_status_dict.drop_duplicates()
display(family_status_dict)

family_status  family_status_id
0         женат / замужем                 0
4        гражданский брак                 1
18         вдовец / вдова                 2
19              в разводе                 3
24  не женат / не замужем                 4

Удалим из исходного датафрейма столбцы education и family_status, оставив только их идентификаторы: education_id и family_status_id. Новые датафреймы — это те самые «словари», к которым мы сможем обращаться по идентификатору.

In [98]:
df = df.drop(['family_status', 'education'], axis=1)

In [99]:
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028       42.0             0                 0      F   
1         1    4024.803754       36.0             1                 0      F   
2         0    5623.422610       33.0             1                 0      M   
3         3    4124.747207       32.0             1                 0      M   
4         0   14177.753002       53.0             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Категоризация дохода

На основании диапазонов, указанных ниже, создадим столбец total_income_category с категориями:
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.
Например, кредитополучателю с доходом 25000 нужно назначить категорию 'E', а клиенту, получающему 235000, — 'B'.

In [100]:
def choose_category(x):
    if 0 < x <= 30_000:
        return 'E'
    if x <= 50_000:
        return 'D'
    if x <= 200_000:
        return 'C'
    if x <= 1_000_000:
        return 'B'
    if x >= 1_000_001:
        return 'A'
    
df['total_income_category'] = df['total_income'].transform(choose_category)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028       42.0             0                 0      F   
1         1    4024.803754       36.0             1                 0      F   
2         0    5623.422610       33.0             1                 0      M   
3         3    4124.747207       32.0             1                 0      M   
4         0   14177.753002       53.0             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Категоризация целей кредита

Изучим данные в столбце purpose и определим, какие подстроки помогут нам правильно определить категорию кредита.

In [101]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
* 'операции с автомобилем' (подстрока 'автомобил'),
* 'операции с недвижимостью' (подстроки 'недвижим', 'жиль')
* 'проведение свадьбы',(подстрока 'свадьб')
* 'получение образования' (подстока 'образован').

Например, если в столбце purpose находится подстрока 'на покупку автомобиля', то в столбце purpose_category должна появиться строка 'операции с автомобилем'.
Используем собственную функцию и метод apply(). 

In [102]:
def label_purpose_category(row):
    if 'автомобил' in row:
        return 'операции с автомобилем'
    if 'недвижим' in row or 'жиль' in row:
        return 'операции с недвижимостью'
    if 'свадьб' in row:
        return 'проведение свадьбы'
    if 'образован' in row:
        return 'получение образования'
    return 'другое'
df['purpose_category'] = df['purpose'].apply(label_purpose_category)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028       42.0             0                 0      F   
1         1    4024.803754       36.0             1                 0      F   
2         0    5623.422610       33.0             1                 0      M   
3         3    4124.747207       32.0             1                 0      M   
4         0   14177.753002       53.0             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

Проверим, что все наши данные корректно классифицированы, категория 'другое' не встречается.

In [103]:
df[df['purpose_category']=='другое']

Empty DataFrame
Columns: [children, days_employed, dob_years, education_id, family_status_id, gender, income_type, debt, total_income, purpose, total_income_category, purpose_category]
Index: []

## Проверка гипотез и ответы на вопросы

Проверим зависимость успешности возврата кредита от ряда характеристик заемщика. 
Сначала определим средний показатель по всем данным, чтобы знать,  с чем сравнивать показатель по конкретным группам.

In [104]:
share = df['debt'].sum()/ df['debt'].count()
print(f'Средний показатель задолженности по кредиту: {share:.2%}')

Средний показатель задолженности по кредиту: 8.12%


Итак, средний показатель задолженности по кредиту составляет 8.12%. 
Проверим, отличается ли он в конкретных группах заемщиков.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

С помощью перекрестных таблиц рассчитаем долю задолженностей у заемщиков с разным количеством детей.

In [107]:
pivot_children = df.pivot_table(index='children', values='debt', aggfunc=['sum','count','mean'])

pivot_children

sum  count      mean
          debt   debt      debt
children                       
0         1063  14091  0.075438
1          445   4855  0.091658
2          202   2128  0.094925
3           27    330  0.081818
4            4     41  0.097561
5            0      9  0.000000

Заемщиков с 4 и 5 детьми мало, для адекватного сравнения по данным группам информации недостаточно. Остальные результаты мы видим - меньше среднего задолженность в группе 'без детей' - 7.55%, в группе с одним и двумя детьми - больше 9%. В то же время родители с тремя детьми не возвращают кредиты в срок примерно в 8% случаев, как и в среднем по всем заемщикам. 

###### Вывод 1:

Наличие детей скорее отрицательно влияет на возврат кредита в срок. Среди клиентов, не имеющих детей доля задолжников меньше, чем среди имеющих детей. 

### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Проверим зависимость между семейным положением и возвратом кредита в срок. Для того, чтобы результаты расчетов были понятнее используем наш словарь family_status_dict, где хранится информация о типах семейного статуса.

In [123]:
pivot_family_status = df.pivot_table(index='family_status_id', values='debt', aggfunc=['sum','count', 'mean'])

pivot_family_status = family_status_dict.merge(pivot_family_status, on='family_status_id', how='left')

#pivot_family_status['Доля задолженностей (%)'] = pivot_family_status[1] / pivot_family_status['All'] * 100
#pivot_family_status['Доля задолженностей (%)'] = pivot_family_status['Доля задолженностей (%)'].round(decimals=2)

pivot_family_status

/home/tanya/anaconda3/envs/practicum/lib/python3.9/site-packages/pandas/core/reshape/merge.py:648: UserWarning: merging between different levels can give an unintended result (1 levels on the left,2 on the right)
  warnings.warn(msg, UserWarning)


family_status  family_status_id  (sum, debt)  (count, debt)  \
0        женат / замужем                 0          931          12339   
1       гражданский брак                 1          388           4151   
2         вдовец / вдова                 2           63            959   
3              в разводе                 3           85           1195   
4  не женат / не замужем                 4          274           2810   

   (mean, debt)  
0      0.075452  
1      0.093471  
2      0.065693  
3      0.071130  
4      0.097509

Все группы достаточно наполнены, поэтому мы можем их все использовать в сравнении.
Меньше всего задолжников среди клиентов с семейным статусом 'вдовец / вдова' - меньше 7%, 'в разводе' - около 7%. Среди находящихся в зарегистрированном браке клиентов-задолжников также меньше среднего - около 7.5%. 
Наиболее 'проблемные' группы заемщиков - среди состоящих в гражданском браке и неженатых клиентов и незамужних клиенток, в обеих категория доля задолженности выше 9%. 

###### Вывод 2:

Клиенты, находящиеся в гражданском браке, и холостые клиенты и неженатые клиенты чаще задерживают платежи по кредитам, чем состоящие в браке и разведенные. Самые же надежные заемщики - вдовцы и вдовы.
На самом деле этот параметр может коррелировать с возрастом - среди клиентов старшего возраста, где чаще представлены вдовцы и вдовы сильнее развита финансовая дисциплина,  а среди молодых людей, среди которых больше свободных и состоящих в гражданском браке может быть распространено менее ответственное финансовое поведение.

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для анализа зависимости между уровнем дохода и возвратом кредита в срок воспользуемся категоризацией клиентов по уровню дохода. Мы разделили клиентов на следующие категории по среднемесячному доходу:
* 0–30000 — 'E'; 
* 30001–50000 — 'D'; 
* 50001–200000 — 'C'; 
* 200001–1000000 — 'B'; 
* 1000001 и выше — 'A'.

In [125]:
pivot_income = df.pivot_table(index='total_income_category',
                              values='debt', 
                              aggfunc=['sum', 'count', 'mean'],
                              margins=True)

pivot_income

sum  count      mean
                       debt   debt      debt
total_income_category                       
A                         2     25  0.080000
B                       356   5042  0.070607
C                      1360  16015  0.084920
D                        21    350  0.060000
E                         2     22  0.090909
All                    1741  21454  0.081150

В категориях А и Е слишком мало случаев для того, чтобы их использовать в межкатегориальных сравнениях.
Малообеспеченные клиенты из категории 'D' возвращают кредит в срок значительно чаще среднего - доля задолжностей около 6%, клиенты с доходом выше среднего из категории 'B' также в среднем чаще возвращают кредиты в срок - в этой категории около 7% задолжностей. Представители среднего класса - категория 'C' чаще допускают просрочку выплат по кредиту.

###### Вывод 3:

Клиенты с доходом выше среднего и ниже среднего допускают просрочки оплаты по кредитам реже, чем представители среднего класса. Это может быть связано с рядом причин: во-первых, среди менее обеспеченных клиентов могут быть пенсионеры, более аккуратные в рассчетах, во-вторых, представители среднего класса активнее других пользуются кредитными инструментами - на них приходится более 3/4 всех выданных кредитов и менее дисциплинированных плательщиков в этой группе так же больше. Они не располагают финансовыми резервами как более обеспеченные категории, но и не располгают дисциплиной как категория 'ниже среднего'.

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [126]:
pivot_purpose = df.pivot_table(index='purpose_category',
                               values='debt',
                               aggfunc=['sum','count','mean'],
                               margins=True)

pivot_purpose

sum  count      mean
                          debt   debt      debt
purpose_category                               
операции с автомобилем     403   4306  0.093590
операции с недвижимостью   782  10811  0.072334
получение образования      370   4013  0.092200
проведение свадьбы         186   2324  0.080034
All                       1741  21454  0.081150

Разные цели кредита влияют на возврат кредита в срок. Операции с недвижимостью - цель кредита, по которой доля задолженности наименьшая- около 7%, на проведение свадьбы - около 8%. Операции с автомобилем и получение образования - цели кредита, по которым доля задолженностей выше среднего - выше 9%.

###### Вывод 4:

Более вероятно, что в срок вернут кредит на операции с недвижимостью, проведение свадьбы. Операции с автомобилем и проведение свадьбы - более 'рискованные' - по ним доля задолженностей выше.

## Общий вывод:

Мы обратились к зависимости между возвратом кредита в срок и такими факторами как количество детей, семейное положение, уровень дохода и цель кредита. 

На возврат кредита в срок отрицательно влияют наличие 1-2 детей, семейный статус 'в гражданском браке', 'не женат / не замужем', средний уровень дохода, цели кредита - 'получение образования' и 'операции с автомобилем'.

Положительно на возврат кредита в срок влияют отсутствие детей, семейный статус 'вдовец / вдова', 'в зарегистрированном браке' или 'в разводе', уровень дохода 'выше среднего' или 'ниже среднего', цель кредита 'операции с недижимостью'. 

Для фактора дохода возможно имеет смысл пересмотреть категоризацию: слишком большой сегмент С можно было бы разбить а несколько групп.

Более точно определить значимость влияния факторов может помочь статистический анализ. Кроме того, полезно было бы проанализировать влияние других факторов, в том числе возраста заемщиков, а также комбинации факторов: семейного положения и наличия детей, возраста и занятости.